src/MQTTClient.h

```diff
@@ -1,63 +1,323 @@
+
+  void setHost(const char hostname[], int port) {
+    this->hostname = hostname;
+    this->port = port;
+  }
+
+  void setWill(const char topic[]) { this->setWill(topic, ""); }
+
+  void setWill(const char topic[], const char payload[]) { this->setWill(topic, payload, false, 0); }
+
+  void setWill(const char topic[], const char payload[], bool retained, int qos) {
+    this->hasWill = true;
+    this->will.topic = lwmqtt_str(topic);
+    this->will.payload = (void *)payload;
+    this->will.payload_len = (int)strlen(payload);
+    this->will.retained = retained;
+    this->will.qos = (lwmqtt_qos_t)qos;
+  }
+
+  void clearWill() { this->hasWill = false; }
+
+  boolean connect(const char clientId[]) { return this->connect(clientId, NULL, NULL); }
+
+  boolean connect(const char clientId[], const char username[], const char password[]) {
+    // return immediately if connected
+    if (this->connected()) {
+      return false;
+    }
+
+    // save client
+    this->network.client = this->netClient;
+
+    // connect to host
+    if (this->netClient->connect(this->hostname, (uint16_t)this->port) < 0) {
+      return false;
+    }
 
-public:
-  MQTTClient();
-  boolean begin(Client &client);
-  boolean begin(const char *hostname, Client &client);
-  boolean begin(const char *hostname, int port, Client &client);
-  void setHost(const char *hostname);
-  void setHost(const char *hostname, int port);
-  void setWill(const char *topic);
-  void setWill(const char *topic, const char *payload);
-  boolean connect(const char *clientId);
-  boolean connect(const char *clientId, const char *username,
-                  const char *password);
-  boolean publish(String topic);
-  boolean publish(String topic, String payload);
-  boolean publish(const char *topic, String payload);
-  boolean publish(const char *topic, const char *payload);
-  boolean publish(const char *topic, char *payload, unsigned int length);
-  boolean publish(MQTTMessage *message);
-  boolean subscribe(String topic);
-  boolean subscribe(const char *topic);
-  boolean unsubscribe(String topic);
-  boolean unsubscribe(const char *topic);
-  void loop();
-  boolean connected();
-  boolean disconnect();
+    // prepare options
+    lwmqtt_options_t options = lwmqtt_default_options;
+    options.client_id = lwmqtt_str(clientId);
+    if (username && password) {
+      options.username = lwmqtt_str(username);
+      options.password = lwmqtt_str(password);
+    }
+
+    // prepare will reference
+    lwmqtt_will_t *will = NULL;
+    if (this->hasWill) {
+      will = &this->will;
+    }
+
+    // connect to broker
+    this->_lastError = lwmqtt_connect(&this->client, &options, will, &this->_returnCode, this->timeout);
+    if (this->_lastError != LWMQTT_SUCCESS) {
+      this->_connected = false;
+      return false;
+    }
+
+    // set flag
+    this->_connected = true;
+
+    return true;
+  }
+
+  boolean publish(String topic) { return this->publish(topic.c_str(), ""); }
```

Comment:
One minor suggestion/optimization is to have all the `String` parameters as `const String&`.